Make DCOS build work on OS X

aaronjwood opened the pull request:
The changes here should make `./build-local.sh` work on OS X. I've updated the `README.md` as well to reflect what is needed on OS X.

I'd like to hear everyone's feedback on this and see if it's something that would be worthwhile to merge.

cmaloney commented:
Looking over briefly this seems like a reasonable approach, and that packages are tagged with the OS they're built in is a good step for now (Will lead to OS X not getting any linux built pacakge caches, but also make sure we don't mix the two, since I don't have confidence in OS X Docker reproducibility of builds yet). Will do an in depth review later.

The code quality failure is a include sorting error:
https://teamcity.mesosphere.io/viewLog.html?buildId=385039&buildTypeId=DcosIo_Dcos_CodeQuality&tab=buildLog&guest=1#_focus=815

You can run the things in the code-quality check by just running `tox`, in this case (for the flake8, pycodestyle, include sorting checks) you can do a `tox -e py34-syntax`.

aaronjwood commented:
Thanks for taking a look! I just fixed the import sorting issue.

cmaloney commented:
I'm going to look internally what it would take for me to setup a basic CI for OS X building (Wouldn't be on every PR because we don't have a fleet of OS X boxes, but should be able to check at least once a day to make sure the build didn't break, and manually trigger on PRs where we want / need)

aaronjwood commented:
Sounds good! One thing I noticed was that the Mesos compilation fails due to running out of memory. I'll add some comments in the `README.md` to indicate that people might need to give their VM a decent amount of memory.

cmaloney commented:
Is there a way we could maybe check if the VM has less than ~5GB of RAM and error early? (We compile mesos with --j8, the biggets files take ~1GB of RAM, and the GCC internal compiler errors when you run out of RAM are terrible for debugging)

aaronjwood commented:
Yeah, I can look into adding a check for that.

aaronjwood commented on build_local.sh:
```diff
@@ -8,6 +8,13 @@ set -o errexit -o pipefail
 
+# Also fail quickly if Docker doesn't have enough memory to build Mesos
+DOCKER_MEMORY=$(docker info | grep Memory | awk '{print $3}')
+if (( $(echo "$DOCKER_MEMORY < 5" | bc -l) )); then
```
@cmaloney do you want to target this to only OS X or keep it as a general check?

cmaloney replied:
I like it in general / everywhere. Linux hosts would also fail with too little RAM in fun mysterious ways (The nodejs / npm build of the UI also uses a ton of RAM)

aaronjwood commented:
I'm running into a strange error when building minuteman:

```
Building package in docker
+ source /opt/mesosphere/environment
++ LD_LIBRARY_PATH=/opt/mesosphere/lib
++ PATH=/opt/mesosphere/bin:/usr/bin:/bin:/sbin
+ pushd /pkg/src/minuteman
/pkg/src/minuteman /
+ make rel
./rebar3 as prod release
===> Fetching pc ({git,"https://github.com/blt/port_compiler.git",
                       {branch,"master"}})
===> Compiling pc
===> Package <<"rebar3_hex">> not found. Fetching registry updates and trying again...
===> Updating package registry...
===> Writing registry to /root/.cache/rebar3/hex/default/registry
===> Generating package index...
===> [cloudi_service_oauth1:1.5.1] Only existing version of cloudi_service_db_riak is 1.3.3 which does not match constraint ~> 1.5.1. Using anyway, but it is not guaranteed to work.
===> Writing index to /root/.cache/rebar3/hex/default/packages.idx
===> Fetching rebar3_hex ({pkg,<<"rebar3_hex">>,<<"2.5.1">>})
===> Downloaded package, caching at /root/.cache/rebar3/hex/default/packages/rebar3_hex-2.5.1.tar
===> Compiling rebar3_hex
===> Linking _build/default/plugins/pc to _build/prod/plugins/pc
===> Linking _build/default/plugins/rebar3_hex to _build/prod/plugins/rebar3_hex
===> Verifying dependencies...
===> Fetching dns ({git,"git://github.com/aetrion/dns_erlang.git",
                        {ref,"674603aec44a2f3fda1f42dd93d1c1368ac319ca"}})
===> Linking _build/default/lib/dns to _build/prod/lib/dns
===> Fetching provider_asn1 ({git,"https://github.com/knusbaum/provider_asn1.git",
                                  {tag,"0.2.0"}})
===> Linking _build/default/plugins/provider_asn1 to _build/prod/plugins/provider_asn1
===> Compiling provider_asn1
===> Fetching eflame ({git,"https://github.com/slfritchie/eflame.git",
                           {ref,"a08518142126f5fc541a3a3c4a04c27f24448bae"}})
===> Linking _build/default/lib/eflame to _build/prod/lib/eflame
===> Fetching eper ({git,"https://github.com/massemanet/eper.git",
                         {ref,"1742031e12fb9e6eac0ffacfc593d5a7849fc07f"}})
===> Linking _build/default/lib/eper to _build/prod/lib/eper
===> Fetching exometer_core ({git,"git://github.com/Feuerlabs/exometer_core",
                                 {ref,
                                     "d7c4ddd066b5a56f7118bdb00dd7b22277340a35"}})
===> Linking _build/default/lib/exometer_core to _build/prod/lib/exometer_core
===> Fetching gen_netlink ({git,"https://github.com/mesosphere/gen_netlink.git",
                                {ref,"e190ef024f7171ddb6d0c5977a1c271e180ed8d4"}})
===> Linking _build/default/lib/gen_netlink to _build/prod/lib/gen_netlink
===> Fetching jsx ({git,"https://github.com/talentdeficit/jsx.git",
                        {ref,"3074d4865b3385a050badf7828ad31490d860df5"}})
===> Linking _build/default/lib/jsx to _build/prod/lib/jsx
===> Fetching lager ({git,"git://github.com/basho/lager",
                          {ref,"8187757388c9adc915379caaab36a2f2ca26e944"}})
===> Linking _build/default/lib/lager to _build/prod/lib/lager
===> Fetching lashup ({git,"git://github.com/dcos/lashup.git",
                           {ref,"a768af34f3230f10eb58baf3a489321278529dee"}})
===> Linking _build/default/lib/lashup to _build/prod/lib/lashup
===> Fetching meck ({git,"git://github.com/eproxus/meck",
                         {ref,"17b0ea5b7d9c93af3a52b5152c6bee14302b3885"}})
===> Linking _build/default/lib/meck to _build/prod/lib/meck
===> Fetching mesos_state ({git,"git://github.com/dcos/mesos_state",
                                {ref,"62992c282fdaebfc52c95d8c9a717068a6b2e199"}})
===> Linking _build/default/lib/mesos_state to _build/prod/lib/mesos_state
===> Fetching rebar3_gpb_plugin ({pkg,<<"rebar3_gpb_plugin">>,<<"1.7.3">>})
===> Downloaded package, caching at /root/.cache/rebar3/hex/default/packages/rebar3_gpb_plugin-1.7.3.tar
===> Linking _build/default/plugins/rebar3_gpb_plugin to _build/prod/plugins/rebar3_gpb_plugin
===> Fetching gpb ({pkg,<<"gpb">>,<<"3.24.4">>})
===> Downloaded package, caching at /root/.cache/rebar3/hex/default/packages/gpb-3.24.4.tar
===> Linking _build/default/plugins/gpb to _build/prod/plugins/gpb
===> Compiling gpb
Compiling descriptor.proto...
Compiling gpb_descriptor.erl...
Compiling gpb_compile_descr.erl...
===> Compiling rebar3_gpb_plugin
===> Fetching pkt ({git,"https://github.com/msantos/pkt",
                        {ref,"e50eb8262922d16ecdbb4b1c1a721480999c6a1a"}})
===> Linking _build/default/lib/pkt to _build/prod/lib/pkt
===> Fetching proper ({git,"https://github.com/manopapad/proper.git",
                           {ref,"1b773eeb47cb2c3116d78bdf681505703b762eee"}})
===> Linking _build/default/lib/proper to _build/prod/lib/proper
===> Fetching sync ({git,"git://github.com/rustyio/sync.git",
                         {ref,"11df81d196eaab2d84caa3fbe8def5d476ef79d8"}})
===> Linking _build/default/lib/sync to _build/prod/lib/sync
===> Fetching telemetry ({git,"git://github.com/dcos/telemetry-net.git",
                              {ref,"1035cf6c59733d52abef7296f0c01fe5ad75e05b"}})
===> Linking _build/default/lib/telemetry to _build/prod/lib/telemetry
===> Fetching webmachine ({git,"git://github.com/webmachine/webmachine",
                               {ref,"b4758b51c05e8f4e8898a70b7a14514a8c150ffa"}})
===> Linking _build/default/lib/webmachine to _build/prod/lib/webmachine
===> Fetching base32 ({git,"https://github.com/aetrion/base32_erlang.git",
                           {ref,"40431a7480176b24863e18b92d9872e54c714df5"}})
===> Linking _build/default/lib/base32 to _build/prod/lib/base32
===> Fetching flower ({git,"https://github.com/travelping/flower.git",
                           {ref,"549015e8ef6e26d6b4bd10c5e527894e55218168"}})
===> Linking _build/default/lib/flower to _build/prod/lib/flower
===> Fetching folsom ({git,"git://github.com/boundary/folsom",
                           {ref,"38e2cce7c64ce1f0a3a918d90394cfc0a940b1ba"}})
===> Linking _build/default/lib/folsom to _build/prod/lib/folsom
===> Fetching gen_socket ({git,"git://github.com/travelping/gen_socket",
                               {ref,"579f41dbf48520c0fbfc9a37217778b834f8e3cb"}})
===> Linking _build/default/lib/gen_socket to _build/prod/lib/gen_socket
===> Fetching goldrush ({git,"git://github.com/DeadZen/goldrush.git",
                             {ref,"212299233c7e7eb63a97be2777e1c05ebaa58dbe"}})
===> Linking _build/default/lib/goldrush to _build/prod/lib/goldrush
===> Fetching gpb ({git,"https://github.com/tomas-abrahamsson/gpb",
                        {ref,"1dbc49f18d6d1979779aff47971dd02405cb509d"}})
===> Linking _build/default/lib/gpb to _build/prod/lib/gpb
===> Fetching mochiweb ({pkg,<<"mochiweb">>,<<"2.12.2">>})
===> Downloaded package, caching at /root/.cache/rebar3/hex/default/packages/mochiweb-2.12.2.tar
===> Linking _build/default/lib/mochiweb to _build/prod/lib/mochiweb
===> Fetching parse_trans ({git,"git://github.com/uwiger/parse_trans.git",
                                {ref,"a210adafdfbb904d156d8f22abd5fb58fc17de1e"}})
===> Linking _build/default/lib/parse_trans to _build/prod/lib/parse_trans
===> Fetching riak_dt ({git,"https://github.com/sargun/riak_dt.git",
                            {ref,"df69cd4a4cb11d7bb97c2ea739c31e614d886bcf"}})
===> Linking _build/default/lib/riak_dt to _build/prod/lib/riak_dt
===> Fetching setup ({git,"git://github.com/uwiger/setup.git",
                          {ref,"85447ca5eb671db0517a475791ac7fb3604bf44d"}})
===> Linking _build/default/lib/setup to _build/prod/lib/setup
===> Fetching bear ({git,"git://github.com/boundary/bear.git",
                         {ref,"119234548783af19b8ec75c879c5062676b92571"}})
===> Linking _build/default/lib/bear to _build/prod/lib/bear
===> Fetching edown ({git,"git://github.com/uwiger/edown.git",
                          {ref,"b7c8eb0ac1859f8fce11cbfe3526f5ec83194776"}})
===> Linking _build/default/lib/edown to _build/prod/lib/edown
===> Compiling goldrush
===> Compiling lager
===> Compiling gen_socket
===> Compiling /pkg/src/minuteman/_build/default/lib/gen_socket/c_src/gen_socket_nif.c
**===> Compiling /pkg/src/minuteman/_build/default/lib/gen_socket/c_src/gen_socket_drv.c
===> Uncaught error in rebar_core. Run with DEBUG=1 to see stacktrace
===> When submitting a bug report, please include the output of `rebar3 report "your command"`
make: *** [rel] Error 1
ERROR Building packages: docker exited non-zero: 2**
Command: docker run -v /Users/user/Code/src/dcos/packages/minuteman/build:/pkg/build:ro -v /Users/user/pkgpanda/pkgpanda_repo8dhozgg6/erlang--caa3176fd2ef42f4f38b5235cad0fc0031c55ff2:/opt/mesosphere/packages/erlang--caa3176fd2ef42f4f38b5235cad0fc0031c55ff2:ro -v /Users/user/Code/src/dcos/packages/cache/packages/minuteman/src:/pkg/src:rw -v /Users/user/pkgpanda/pkgpanda-nzvdc_ik:/opt/mesosphere:ro -v /Users/user/pkgpanda/pkgpanda_repo8dhozgg6/ncurses--f6bab4377341cc166964b4756ace43c3ad71fac9:/opt/mesosphere/packages/ncurses--f6bab4377341cc166964b4756ace43c3ad71fac9:ro -v /Users/user/pkgpanda/pkgpanda_repo8dhozgg6/openssl--014004a5e4da0c8a1f853865945648dfa6cbe693:/opt/mesosphere/packages/openssl--014004a5e4da0c8a1f853865945648dfa6cbe693:ro -v /Users/user/Code/src/dcos/packages/cache/packages/minuteman/result:/opt/mesosphere/packages/minuteman--adb4a2fdce9404bf05450d1be140fe1630fe34af:rw -e OS=darwin -e PKG_PATH=/opt/mesosphere/packages/minuteman--adb4a2fdce9404bf05450d1be140fe1630fe34af -e PKG_ID=minuteman--adb4a2fdce9404bf05450d1be140fe1630fe34af -e PKG_NAME=minuteman -e PKG_VERSION=adb4a2fdce9404bf05450d1be140fe1630fe34af -e PKG_VARIANT=<default> dcos/dcos-builder:dcos-builder_dockerdir-latest /bin/bash -o nounset -o pipefail -o errexit /pkg/build
Traceback (most recent call last):
  File "/Users/user/Code/src/dcos/pkgpanda/build/__init__.py", line 1183, in build
    "/pkg/build"])
  File "/Users/user/Code/src/dcos/pkgpanda/build/__init__.py", line 51, in run
    check_call(docker)
  File "/usr/local/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/subprocess.py", line 581, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['docker', 'run', '-v', '/Users/user/Code/src/dcos/packages/minuteman/build:/pkg/build:ro', '-v', '/Users/user/pkgpanda/pkgpanda_repo8dhozgg6/erlang--caa3176fd2ef42f4f38b5235cad0fc0031c55ff2:/opt/mesosphere/packages/erlang--caa3176fd2ef42f4f38b5235cad0fc0031c55ff2:ro', '-v', '/Users/user/Code/src/dcos/packages/cache/packages/minuteman/src:/pkg/src:rw', '-v', '/Users/user/pkgpanda/pkgpanda-nzvdc_ik:/opt/mesosphere:ro', '-v', '/Users/user/pkgpanda/pkgpanda_repo8dhozgg6/ncurses--f6bab4377341cc166964b4756ace43c3ad71fac9:/opt/mesosphere/packages/ncurses--f6bab4377341cc166964b4756ace43c3ad71fac9:ro', '-v', '/Users/user/pkgpanda/pkgpanda_repo8dhozgg6/openssl--014004a5e4da0c8a1f853865945648dfa6cbe693:/opt/mesosphere/packages/openssl--014004a5e4da0c8a1f853865945648dfa6cbe693:ro', '-v', '/Users/user/Code/src/dcos/packages/cache/packages/minuteman/result:/opt/mesosphere/packages/minuteman--adb4a2fdce9404bf05450d1be140fe1630fe34af:rw', '-e', 'OS=darwin', '-e', 'PKG_PATH=/opt/mesosphere/packages/minuteman--adb4a2fdce9404bf05450d1be140fe1630fe34af', '-e', 'PKG_ID=minuteman--adb4a2fdce9404bf05450d1be140fe1630fe34af', '-e', 'PKG_NAME=minuteman', '-e', 'PKG_VERSION=adb4a2fdce9404bf05450d1be140fe1630fe34af', '-e', 'PKG_VARIANT=<default>', 'dcos/dcos-builder:dcos-builder_dockerdir-latest', '/bin/bash', '-o', 'nounset', '-o', 'pipefail', '-o', 'errexit', '/pkg/build']' returned non-zero exit status 2

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/dcos_build_venv/bin/release", line 9, in <module>
    load_entry_point('dcos-image', 'console_scripts', 'release')()
  File "/Users/user/Code/src/dcos/release/__init__.py", line 783, in main
    release_manager.create('testing', options.channel, options.tag)
  File "/Users/user/Code/src/dcos/release/__init__.py", line 676, in create
    metadata = make_stable_artifacts(self.__config['options']['cloudformation_s3_url'] + '/' + repository_path)
  File "/Users/user/Code/src/dcos/release/__init__.py", line 280, in make_stable_artifacts
    all_completes = do_build_packages(cache_repository_url)
  File "/Users/user/Code/src/dcos/release/__init__.py", line 494, in do_build_packages
    result = pkgpanda.build.build_tree(package_store, True, None)
  File "/Users/user/Code/src/dcos/pkgpanda/build/__init__.py", line 670, in build_tree
    True)
  File "/Users/user/Code/src/dcos/pkgpanda/build/__init__.py", line 1185, in build
    raise BuildError("docker exited non-zero: {}\nCommand: {}".format(ex.returncode, ' '.join(ex.cmd)))
pkgpanda.build.BuildError: docker exited non-zero: 2
Command: docker run -v /Users/user/Code/src/dcos/packages/minuteman/build:/pkg/build:ro -v /Users/user/pkgpanda/pkgpanda_repo8dhozgg6/erlang--caa3176fd2ef42f4f38b5235cad0fc0031c55ff2:/opt/mesosphere/packages/erlang--caa3176fd2ef42f4f38b5235cad0fc0031c55ff2:ro -v /Users/user/Code/src/dcos/packages/cache/packages/minuteman/src:/pkg/src:rw -v /Users/user/pkgpanda/pkgpanda-nzvdc_ik:/opt/mesosphere:ro -v /Users/user/pkgpanda/pkgpanda_repo8dhozgg6/ncurses--f6bab4377341cc166964b4756ace43c3ad71fac9:/opt/mesosphere/packages/ncurses--f6bab4377341cc166964b4756ace43c3ad71fac9:ro -v /Users/user/pkgpanda/pkgpanda_repo8dhozgg6/openssl--014004a5e4da0c8a1f853865945648dfa6cbe693:/opt/mesosphere/packages/openssl--014004a5e4da0c8a1f853865945648dfa6cbe693:ro -v /Users/user/Code/src/dcos/packages/cache/packages/minuteman/result:/opt/mesosphere/packages/minuteman--adb4a2fdce9404bf05450d1be140fe1630fe34af:rw -e OS=darwin -e PKG_PATH=/opt/mesosphere/packages/minuteman--adb4a2fdce9404bf05450d1be140fe1630fe34af -e PKG_ID=minuteman--adb4a2fdce9404bf05450d1be140fe1630fe34af -e PKG_NAME=minuteman -e PKG_VERSION=adb4a2fdce9404bf05450d1be140fe1630fe34af -e PKG_VARIANT=<default> dcos/dcos-builder:dcos-builder_dockerdir-latest /bin/bash -o nounset -o pipefail -o errexit /pkg/build
```

Doesn't seem to be related to memory...

cmaloney commented:
Pinged the engineers who work on minuteman

cmaloney commented:
Could you update the build file in the minuteman package, to do a `export DEBUG=1` around: https://github.com/dcos/dcos/blob/master/packages/minuteman/build#L5

then try rebuilding it and post the new error?

aaronjwood commented:
Sure, there's a lot of output so here's the section where it starts to fail:

```
===> Compiling /pkg/src/minuteman/_build/default/lib/gen_socket/c_src/gen_socket_drv.c
===> /pkg/src/minuteman/_build/default/lib/gen_socket/c_src/gen_socket_drv.c: In function 'gs_send_single_atom':
/pkg/src/minuteman/_build/default/lib/gen_socket/c_src/gen_socket_drv.c:163:2: warning: 'driver_output_term' is deprecated (declared at /opt/mesosphere/packages/erlang--caa3176fd2ef42f4f38b5235cad0fc0031c55ff2/lib/erlang/erts-8.0/include/erl_driver.h:600) [-Wdeprecated-declarations]
  driver_output_term(state->drv_port, output, sizeof(output) / sizeof(ErlDrvTermData));
  ^

===> Uncaught error in rebar_core. Run with DEBUG=1 to see stacktrace
===> Uncaught error: function_clause
===> Stack trace to the error location: [{lists,reverse,
                                          [["/pkg/src/minuteman/_build/default/lib/gen_socket/c_src/gen_socket_drv.o"|
                                            {["/pkg/src/minuteman/_build/default/lib/gen_socket/c_src/gen_socket_nif.o"],
                                             ["{ \"file\"      : \"/pkg/src/minuteman/_build/default/lib/gen_socket/c_src/gen_socket_nif.c\"\n, \"directory\" : \"/pkg/src/minuteman/_build/default/lib/gen_socket\"\n, \"command\"   : \"cc -c -g -Wall -fPIC -MMD -I\\\"/opt/mesosphere/packages/erlang--caa3176fd2ef42f4f38b5235cad0fc0031c55ff2/lib/erlang/lib/erl_interface-3.9/include\\\" -I\\\"/opt/mesosphere/packages/erlang--caa3176fd2ef42f4f38b5235cad0fc0031c55ff2/lib/erlang/erts-8.0/include\\\" /pkg/src/minuteman/_build/default/lib/gen_socket/c_src/gen_socket_nif.c -o /pkg/src/minuteman/_build/default/lib/gen_socket/c_src/gen_socket_nif.o\"\n}~n"]}]],
                                          [{file,"lists.erl"},{line,147}]},
                                         {pc_compilation,compile_each,5,
                                          [{file,
                                            "/pkg/src/minuteman/_build/default/plugins/pc/src/pc_compilation.erl"},
                                           {line,117}]},
                                         {pc_compilation,
                                          '-compile_sources/2-lists^foldl/2-0-',
                                          3,
                                          [{file,
                                            "/pkg/src/minuteman/_build/default/plugins/pc/src/pc_compilation.erl"},
                                           {line,95}]},
                                         {pc_compilation,compile_sources,2,
                                          [{file,
                                            "/pkg/src/minuteman/_build/default/plugins/pc/src/pc_compilation.erl"},
                                           {line,95}]},
                                         {pc_compilation,compile_and_link,2,
                                          [{file,
                                            "/pkg/src/minuteman/_build/default/plugins/pc/src/pc_compilation.erl"},
                                           {line,42}]},
                                         {pc_prv_compile,'-do/1-lc$^0/1-0-',
                                          2,
                                          [{file,
                                            "/pkg/src/minuteman/_build/default/plugins/pc/src/pc_prv_compile.erl"},
                                           {line,53}]},
                                         {pc_prv_compile,do,1,
                                          [{file,
                                            "/pkg/src/minuteman/_build/default/plugins/pc/src/pc_prv_compile.erl"},
                                           {line,49}]},
                                         {rebar_core,do,2,
                                          [{file,
                                            "/home/tristan/Devel/rebar3/_build/default/lib/rebar/src/rebar_core.erl"},
                                           {line,125}]}]
===> When submitting a bug report, please include the output of `rebar3 report "your command"`
make: *** [rel] Error 1
ERROR Building packages: docker exited non-zero: 2
Command: docker run -v /Users/user/pkgpanda/pkgpanda_repoxhpi5giz/erlang--caa3176fd2ef42f4f38b5235cad0fc0031c55ff2:/opt/mesosphere/packages/erlang--caa3176fd2ef42f4f38b5235cad0fc0031c55ff2:ro -v /Users/user/Code/src/dcos/packages/minuteman/build:/pkg/build:ro -v /Users/user/Code/src/dcos/packages/cache/packages/minuteman/src:/pkg/src:rw -v /Users/user/pkgpanda/pkgpanda_repoxhpi5giz/ncurses--f6bab4377341cc166964b4756ace43c3ad71fac9:/opt/mesosphere/packages/ncurses--f6bab4377341cc166964b4756ace43c3ad71fac9:ro -v /Users/user/Code/src/dcos/packages/cache/packages/minuteman/result:/opt/mesosphere/packages/minuteman--cd1d326fb908e61855cb056a48e83078f5078ef0:rw -v /Users/user/pkgpanda/pkgpanda_repoxhpi5giz/openssl--014004a5e4da0c8a1f853865945648dfa6cbe693:/opt/mesosphere/packages/openssl--014004a5e4da0c8a1f853865945648dfa6cbe693:ro -v /Users/user/pkgpanda/pkgpanda-pm_38edd:/opt/mesosphere:ro -e PKG_VERSION=cd1d326fb908e61855cb056a48e83078f5078ef0 -e PKG_NAME=minuteman -e PKG_VARIANT=<default> -e PKG_ID=minuteman--cd1d326fb908e61855cb056a48e83078f5078ef0 -e PKG_PATH=/opt/mesosphere/packages/minuteman--cd1d326fb908e61855cb056a48e83078f5078ef0 -e OS=darwin dcos/dcos-builder:dcos-builder_dockerdir-latest /bin/bash -o nounset -o pipefail -o errexit /pkg/build
Traceback (most recent call last):
  File "/Users/user/Code/src/dcos/pkgpanda/build/__init__.py", line 1183, in build
    "/pkg/build"])
  File "/Users/user/Code/src/dcos/pkgpanda/build/__init__.py", line 51, in run
    check_call(docker)
  File "/usr/local/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/subprocess.py", line 581, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['docker', 'run', '-v', '/Users/user/pkgpanda/pkgpanda_repoxhpi5giz/erlang--caa3176fd2ef42f4f38b5235cad0fc0031c55ff2:/opt/mesosphere/packages/erlang--caa3176fd2ef42f4f38b5235cad0fc0031c55ff2:ro', '-v', '/Users/user/Code/src/dcos/packages/minuteman/build:/pkg/build:ro', '-v', '/Users/user/Code/src/dcos/packages/cache/packages/minuteman/src:/pkg/src:rw', '-v', '/Users/user/pkgpanda/pkgpanda_repoxhpi5giz/ncurses--f6bab4377341cc166964b4756ace43c3ad71fac9:/opt/mesosphere/packages/ncurses--f6bab4377341cc166964b4756ace43c3ad71fac9:ro', '-v', '/Users/user/Code/src/dcos/packages/cache/packages/minuteman/result:/opt/mesosphere/packages/minuteman--cd1d326fb908e61855cb056a48e83078f5078ef0:rw', '-v', '/Users/user/pkgpanda/pkgpanda_repoxhpi5giz/openssl--014004a5e4da0c8a1f853865945648dfa6cbe693:/opt/mesosphere/packages/openssl--014004a5e4da0c8a1f853865945648dfa6cbe693:ro', '-v', '/Users/user/pkgpanda/pkgpanda-pm_38edd:/opt/mesosphere:ro', '-e', 'PKG_VERSION=cd1d326fb908e61855cb056a48e83078f5078ef0', '-e', 'PKG_NAME=minuteman', '-e', 'PKG_VARIANT=<default>', '-e', 'PKG_ID=minuteman--cd1d326fb908e61855cb056a48e83078f5078ef0', '-e', 'PKG_PATH=/opt/mesosphere/packages/minuteman--cd1d326fb908e61855cb056a48e83078f5078ef0', '-e', 'OS=darwin', 'dcos/dcos-builder:dcos-builder_dockerdir-latest', '/bin/bash', '-o', 'nounset', '-o', 'pipefail', '-o', 'errexit', '/pkg/build']' returned non-zero exit status 2

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/dcos_build_venv/bin/release", line 9, in <module>
    load_entry_point('dcos-image', 'console_scripts', 'release')()
  File "/Users/user/Code/src/dcos/release/__init__.py", line 783, in main
    release_manager.create('testing', options.channel, options.tag)
  File "/Users/user/Code/src/dcos/release/__init__.py", line 676, in create
    metadata = make_stable_artifacts(self.__config['options']['cloudformation_s3_url'] + '/' + repository_path)
  File "/Users/user/Code/src/dcos/release/__init__.py", line 280, in make_stable_artifacts
    all_completes = do_build_packages(cache_repository_url)
  File "/Users/user/Code/src/dcos/release/__init__.py", line 494, in do_build_packages
    result = pkgpanda.build.build_tree(package_store, True, None)
  File "/Users/user/Code/src/dcos/pkgpanda/build/__init__.py", line 670, in build_tree
    True)
  File "/Users/user/Code/src/dcos/pkgpanda/build/__init__.py", line 1185, in build
    raise BuildError("docker exited non-zero: {}\nCommand: {}".format(ex.returncode, ' '.join(ex.cmd)))
pkgpanda.build.BuildError: docker exited non-zero: 2
Command: docker run -v /Users/user/pkgpanda/pkgpanda_repoxhpi5giz/erlang--caa3176fd2ef42f4f38b5235cad0fc0031c55ff2:/opt/mesosphere/packages/erlang--caa3176fd2ef42f4f38b5235cad0fc0031c55ff2:ro -v /Users/user/Code/src/dcos/packages/minuteman/build:/pkg/build:ro -v /Users/user/Code/src/dcos/packages/cache/packages/minuteman/src:/pkg/src:rw -v /Users/user/pkgpanda/pkgpanda_repoxhpi5giz/ncurses--f6bab4377341cc166964b4756ace43c3ad71fac9:/opt/mesosphere/packages/ncurses--f6bab4377341cc166964b4756ace43c3ad71fac9:ro -v /Users/user/Code/src/dcos/packages/cache/packages/minuteman/result:/opt/mesosphere/packages/minuteman--cd1d326fb908e61855cb056a48e83078f5078ef0:rw -v /Users/user/pkgpanda/pkgpanda_repoxhpi5giz/openssl--014004a5e4da0c8a1f853865945648dfa6cbe693:/opt/mesosphere/packages/openssl--014004a5e4da0c8a1f853865945648dfa6cbe693:ro -v /Users/user/pkgpanda/pkgpanda-pm_38edd:/opt/mesosphere:ro -e PKG_VERSION=cd1d326fb908e61855cb056a48e83078f5078ef0 -e PKG_NAME=minuteman -e PKG_VARIANT=<default> -e PKG_ID=minuteman--cd1d326fb908e61855cb056a48e83078f5078ef0 -e PKG_PATH=/opt/mesosphere/packages/minuteman--cd1d326fb908e61855cb056a48e83078f5078ef0 -e OS=darwin dcos/dcos-builder:dcos-builder_dockerdir-latest /bin/bash -o nounset -o pipefail -o errexit /pkg/build
```

tsloughter commented:
@aaronjwood can you try again fresh?

Also, I'd suggest that the `pc` plugin in minuteman be switched to use `{pc, "1.3.0"}` so it is fetched from hex and not master.

aaronjwood commented:
@tsloughter I removed the cached minuteman package and kicked off the build again. Looks like the issue is fixed now. Thanks!

aaronjwood commented:
Now navstar isn't working :)

```
Cloning into bare repository '/Users/user/Code/src/dcos/packages/cache/packages/navstar/navstar/cache.git'...
remote: Counting objects: 575, done.
remote: Compressing objects: 100% (57/57), done.
remote: Total 575 (delta 25), reused 0 (delta 0), pack-reused 518
Receiving objects: 100% (575/575), 1.93 MiB | 1.70 MiB/s, done.
Resolving deltas: 100% (326/326), done.
Checking connectivity... done.
WARNING: Current ref doesn't match the ref origin. Package ref should probably be updated to pick up new changes to the code: Current: aaa2e6609ec2511743ab1bf58876893eb4f4a654, Origin: 2bc3a6ff71d461743dbc30a79521104262e9a986
fatal: reference is not a tree: aaa2e6609ec2511743ab1bf58876893eb4f4a654
Traceback (most recent call last):
  File "/tmp/dcos_build_venv/bin/release", line 9, in <module>
    load_entry_point('dcos-image', 'console_scripts', 'release')()
  File "/Users/user/Code/src/dcos/release/__init__.py", line 783, in main
    release_manager.create('testing', options.channel, options.tag)
  File "/Users/user/Code/src/dcos/release/__init__.py", line 676, in create
    metadata = make_stable_artifacts(self.__config['options']['cloudformation_s3_url'] + '/' + repository_path)
  File "/Users/user/Code/src/dcos/release/__init__.py", line 280, in make_stable_artifacts
    all_completes = do_build_packages(cache_repository_url)
  File "/Users/user/Code/src/dcos/release/__init__.py", line 494, in do_build_packages
    result = pkgpanda.build.build_tree(package_store, True, None)
  File "/Users/user/Code/src/dcos/pkgpanda/build/__init__.py", line 670, in build_tree
    True)
  File "/Users/user/Code/src/dcos/pkgpanda/build/__init__.py", line 1108, in build
    fetcher.checkout_to(root)
  File "/Users/user/Code/src/dcos/pkgpanda/build/src_fetchers.py", line 126, in checkout_to
    self.ref])
  File "/usr/local/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/subprocess.py", line 581, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['git', '--git-dir', '/Users/user/Code/src/dcos/packages/cache/packages/navstar/src/navstar/.git', '--work-tree', '/Users/user/Code/src/dcos/packages/cache/packages/navstar/src/navstar', 'checkout', '-f', '-q', 'aaa2e6609ec2511743ab1bf58876893eb4f4a654']' returned non-zero exit status 128
```

cmaloney commented:
@aaronjwood can you rebase on master? https://github.com/dcos/dcos/pull/582 updated the commit referenced there

As far as why the old commit disappeared, I'm not sure, pinging the team working on that bit.

aaronjwood commented:
That fixed it, thanks!

aaronjwood commented:
Just made it all the way through to the end. Looks like I have all of the artifacts that I should.

aaronjwood commented on prep_local:
```diff
@@ -10,7 +10,7 @@ set -o errexit -o nounset -o pipefail
 
 # Install the latest version of pip
-pip install pip==8.1.2
+pip install --upgrade pip
```
@cmaloney I found that without this change you'll see errors like this in 16.04, Mint 18, etc.

```
+ ./prep_local
Exception:
Traceback (most recent call last):
  File "/tmp/dcos_build_venv/lib/python3.5/site-packages/pip/req/req_install.py", line 1006, in check_if_exists
    self.satisfied_by = pkg_resources.get_distribution(str(no_marker))
  File "/tmp/dcos_build_venv/share/python-wheels/pkg_resources-0.0.0-py2.py3-none-any.whl/pkg_resources/__init__.py", line 535, in get_distribution
    dist = get_provider(dist)
  File "/tmp/dcos_build_venv/share/python-wheels/pkg_resources-0.0.0-py2.py3-none-any.whl/pkg_resources/__init__.py", line 415, in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
  File "/tmp/dcos_build_venv/share/python-wheels/pkg_resources-0.0.0-py2.py3-none-any.whl/pkg_resources/__init__.py", line 695, in find
    raise VersionConflict(dist, req)
pkg_resources.VersionConflict: (pip 8.1.1 (/tmp/dcos_build_venv/lib/python3.5/site-packages), Requirement.parse('pip==8.1.2'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/dcos_build_venv/lib/python3.5/site-packages/pip/basecommand.py", line 209, in main
    status = self.run(options, args)
  File "/tmp/dcos_build_venv/lib/python3.5/site-packages/pip/commands/install.py", line 328, in run
    wb.build(autobuilding=True)
  File "/tmp/dcos_build_venv/lib/python3.5/site-packages/pip/wheel.py", line 748, in build
    self.requirement_set.prepare_files(self.finder)
  File "/tmp/dcos_build_venv/lib/python3.5/site-packages/pip/req/req_set.py", line 360, in prepare_files
    ignore_dependencies=self.ignore_dependencies))
  File "/tmp/dcos_build_venv/lib/python3.5/site-packages/pip/req/req_set.py", line 448, in _prepare_file
    req_to_install, finder)
  File "/tmp/dcos_build_venv/lib/python3.5/site-packages/pip/req/req_set.py", line 387, in _check_skip_installed
    req_to_install.check_if_exists()
  File "/tmp/dcos_build_venv/lib/python3.5/site-packages/pip/req/req_install.py", line 1011, in check_if_exists
    self.req.project_name
AttributeError: 'Requirement' object has no attribute 'project_name'
You are using pip version 8.1.1, however version 8.1.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```

orsenthil replied:
This is just looks like an ugly "pip" issue, that got fixed from 8.1.1 to 8.1.2

https://github.com/pypa/pip/commit/8e236dd6a09bd2f70f9d4fc886da8c354d4c58f2#diff-297395da5417032dc5d0738a1fc7e136L1002

For pip itself, I think.

`pip install --upgrade pip`

is a reasonable solution.

For other packages, we will require the hardcoded version.

So, this change looks good to me.

If you want to be conservative, remove the `pip install pip==8.1.2` line itself and we can upgrade pip in the buildinfo.json.  
Since this is a bug in pip, these are the two only two possible options for us.

cmaloney replied:
Lets move this and the package to 8.1.2 then. Really want to keep them in sync.

aaronjwood replied:
So do you guys want me to revert this change and just make the change to the version in `buildinfo.json`?

orsenthil replied:
I will revert it in a separate PR. Let's not confuse it with this one.  thank you.

cmaloney commented:
The pinning to "exactly" 8.1.2 is actually intentional since we also have some internal stuff which uses that precise version: https://github.com/dcos/dcos/blob/master/packages/python/buildinfo.json#L20

cmaloney commented:
@orsenthil any insights?

aaronjwood commented:
I can revert the change, any other ideas how to get around the failures?

cmaloney commented:
That's precisely what I'm hoping senthil can help with, he knows more of that area of internal python bits than I do. In theory upgrading the pip in that particular virtualenv should be fine, but it's failing which is odd...

orsenthil commented on packages/python/build:
```diff
@@ -18,6 +18,12 @@ rm -rf "$PKG_PATH/lib/python3.4/test"
+# This is due to the build being performed inside of the VM's shared folder.
+# https://www.virtualbox.org/ticket/818
+if [[ "$OS" == "darwin" ]]; then
+    cp -f "$PKG_PATH/bin/python3.4m" "$PKG_PATH/bin/python3.4"
```
I ran into a problem with this change.

```
cp: '/opt/mesosphere/packages/python--2de80c630ae82c807c0a63d24a6eed8640faf2f6/bin/python3.4m' and '/opt/mesosphere/packages/python--2de80c630ae82c807c0a63d24a6eed8640faf2f6/bin/python3.4' are the same file
ERROR Building packages: docker exited non-zero: 1
Command: docker run -v /Users/senthil/workspace/dcos/dcos/packages/cache/packages/python/src:/pkg/src:rw -v /Users/senthil/workspace/dcos/dcos/packages/python/build:/pkg/build:ro -v /Users/senthil/workspace/dcos/dcos/packages/cache/packages/python/result:/opt/mesosphere/packages/python--2de80c630ae82c807c0a63d24a6eed8640faf2f6:rw -v /Users/senthil/pkgpanda/pkgpanda-ba9bfkxu:/opt/mesosphere:ro -v /Users/senthil/pkgpanda/pkgpanda_repor5ozy_h4/openssl--b549a917aa074e6fa349fbb15c405d074b3d3b8c:/opt/mesosphere/packages/openssl--b549a917aa074e6fa349fbb15c405d074b3d3b8c:ro -e PKG_VERSION=2de80c630ae82c807c0a63d24a6eed8640faf2f6 -e PKG_NAME=python -e PKG_PATH=/opt/mesosphere/packages/python--2de80c630ae82c807c0a63d24a6eed8640faf2f6 -e OS=darwin -e PKG_VARIANT=<default> -e PKG_ID=python--2de80c630ae82c807c0a63d24a6eed8640faf2f6 dcos/dcos-builder:dcos-builder_dockerdir-latest /bin/bash -o nounset -o pipefail -o errexit /pkg/build
Traceback (most recent call last):
  File "/Users/senthil/workspace/dcos/dcos/pkgpanda/build/__init__.py", line 1184, in build
    "/pkg/build"])
  File "/Users/senthil/workspace/dcos/dcos/pkgpanda/build/__init__.py", line 51, in run
    check_call(docker)
  File "/Users/senthil/python3.4/lib/python3.4/subprocess.py", line 558, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['docker', 'run', '-v', '/Users/senthil/workspace/dcos/dcos/packages/cache/packages/python/src:/pkg/src:rw', '-v', '/Users/senthil/workspace/dcos/dcos/packages/python/build:/pkg/build:ro', '-v', '/Users/senthil/workspace/dcos/dcos/packages/cache/packages/python/result:/opt/mesosphere/packages/python--2de80c630ae82c807c0a63d24a6eed8640faf2f6:rw', '-v', '/Users/senthil/pkgpanda/pkgpanda-ba9bfkxu:/opt/mesosphere:ro', '-v', '/Users/senthil/pkgpanda/pkgpanda_repor5ozy_h4/openssl--b549a917aa074e6fa349fbb15c405d074b3d3b8c:/opt/mesosphere/packages/openssl--b549a917aa074e6fa349fbb15c405d074b3d3b8c:ro', '-e', 'PKG_VERSION=2de80c630ae82c807c0a63d24a6eed8640faf2f6', '-e', 'PKG_NAME=python', '-e', 'PKG_PATH=/opt/mesosphere/packages/python--2de80c630ae82c807c0a63d24a6eed8640faf2f6', '-e', 'OS=darwin', '-e', 'PKG_VARIANT=<default>', '-e', 'PKG_ID=python--2de80c630ae82c807c0a63d24a6eed8640faf2f6', 'dcos/dcos-builder:dcos-builder_dockerdir-latest', '/bin/bash', '-o', 'nounset', '-o', 'pipefail', '-o', 'errexit', '/pkg/build']' returned non-zero exit status 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/dcos_build_venv/bin/release", line 9, in <module>
    load_entry_point('dcos-image', 'console_scripts', 'release')()
  File "/Users/senthil/workspace/dcos/dcos/release/__init__.py", line 813, in main
    release_manager.create('testing', options.channel, options.tag)
  File "/Users/senthil/workspace/dcos/dcos/release/__init__.py", line 718, in create
    metadata = make_stable_artifacts(self.__config['options']['cloudformation_s3_url'] + '/' + repository_path)
  File "/Users/senthil/workspace/dcos/dcos/release/__init__.py", line 302, in make_stable_artifacts
    all_completes = do_build_packages(cache_repository_url)
  File "/Users/senthil/workspace/dcos/dcos/release/__init__.py", line 519, in do_build_packages
    result = pkgpanda.build.build_tree(package_store, True, None)
  File "/Users/senthil/workspace/dcos/dcos/pkgpanda/build/__init__.py", line 670, in build_tree
    True)
  File "/Users/senthil/workspace/dcos/dcos/pkgpanda/build/__init__.py", line 1186, in build
    raise BuildError("docker exited non-zero: {}\nCommand: {}".format(ex.returncode, ' '.join(ex.cmd)))
pkgpanda.build.BuildError: docker exited non-zero: 1
Command: docker run -v /Users/senthil/workspace/dcos/dcos/packages/cache/packages/python/src:/pkg/src:rw -v /Users/senthil/workspace/dcos/dcos/packages/python/build:/pkg/build:ro -v /Users/senthil/workspace/dcos/dcos/packages/cache/packages/python/result:/opt/mesosphere/packages/python--2de80c630ae82c807c0a63d24a6eed8640faf2f6:rw -v /Users/senthil/pkgpanda/pkgpanda-ba9bfkxu:/opt/mesosphere:ro -v /Users/senthil/pkgpanda/pkgpanda_repor5ozy_h4/openssl--b549a917aa074e6fa349fbb15c405d074b3d3b8c:/opt/mesosphere/packages/openssl--b549a917aa074e6fa349fbb15c405d074b3d3b8c:ro -e PKG_VERSION=2de80c630ae82c807c0a63d24a6eed8640faf2f6 -e PKG_NAME=python -e PKG_PATH=/opt/mesosphere/packages/python--2de80c630ae82c807c0a63d24a6eed8640faf2f6 -e OS=darwin -e PKG_VARIANT=<default> -e PKG_ID=python--2de80c630ae82c807c0a63d24a6eed8640faf2f6 dcos/dcos-builder:dcos-builder_dockerdir-latest /bin/bash -o nounset -o pipefail -o errexit /pkg/build
/Users/senthil/python3.4/lib/python3.4/tempfile.py:699: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/Users/senthil/pkgpanda/pkgpanda_repor5ozy_h4'>
  _warnings.warn(warn_message, ResourceWarning)

```

I had to revert this portion of change to proceed.

aaronjwood replied:
Strange, how did you hit that on OS X? I am unable to reproduce your error. Although it maybe be because I'm building with Docker Machine and I'm assuming that the hard link won't work. What I can do is make a copy to some temporary name and then swap the file instead. What do you think?

orsenthil replied:
I am using Docker on Mac and I had not dependency on Virtualbox (Actually the comment referencing virtualbox surprised me a bit).

* For OSX, I think, it is safe to assume that customers will be using "Docker on Mac", and if we target that, then this `code-stanza` is not required.

aaronjwood replied:
Okay, if that's what we want to target I will go ahead and revert this change.

orsenthil commented:
(sorry unable to comment inline) On ` tar = "gtar" if sys.platform == "darwin" else "tar"`  Ensure that the binary exists before proceeding. if the binary did not exist, bail out. This is similar checking if docker has been allocated more than 5GB Ram.

Most often folks miss the fine nuances of README.

orsenthil commented:
(sorry unable to comment inline) On ` tar = "gtar" if sys.platform == "darwin" else "tar"`  Ensure that the binary exists before proceeding. if the binary did not exist, bail out. This is similar checking if docker has been allocated more than 5GB Ram.

Most often folks miss the fine nuances of README.

aaronjwood commented:
@orsenthil sure, I can add this check in. I'll put it in the same script where I added the RAM check for Docker, sound good?

aaronjwood commented:
@orsenthil just added the check for `gtar`.

aaronjwood commented:
This is unrelated but I thought I'd mention that the build is failing (this was on Linux):

```
build	07-Sep-2016 12:51:16	Validating CloudFormation: aws/templates/advanced/infra.json
error	07-Sep-2016 12:51:19	Traceback (most recent call last):
error	07-Sep-2016 12:51:19	  File "/tmp/dcos_build_venv/bin/release", line 9, in <module>
error	07-Sep-2016 12:51:19	    load_entry_point('dcos-image==0.1', 'console_scripts', 'release')()
error	07-Sep-2016 12:51:19	  File "/home/jenkins/release/__init__.py", line 813, in main
error	07-Sep-2016 12:51:19	    release_manager.create('testing', options.channel, options.tag)
error	07-Sep-2016 12:51:19	  File "/home/jenkins/release/__init__.py", line 729, in create
error	07-Sep-2016 12:51:19	    metadata['channel_artifacts'] = make_channel_artifacts(metadata)
error	07-Sep-2016 12:51:19	  File "/home/jenkins/release/__init__.py", line 407, in make_channel_artifacts
error	07-Sep-2016 12:51:19	    all_bootstraps=metadata["all_bootstraps"]):
error	07-Sep-2016 12:51:19	  File "/home/jenkins/gen/installer/azure.py", line 234, in do_create
error	07-Sep-2016 12:51:19	    'local_content': gen_buttons(build_name, reproducible_artifact_path, tag, commit),
error	07-Sep-2016 12:51:19	  File "/home/jenkins/gen/installer/azure.py", line 248, in gen_buttons
error	07-Sep-2016 12:51:19	    for x in [1, 3, 5]]
error	07-Sep-2016 12:51:19	  File "/home/jenkins/gen/installer/azure.py", line 248, in <listcomp>
error	07-Sep-2016 12:51:19	    for x in [1, 3, 5]]
error	07-Sep-2016 12:51:19	  File "/home/jenkins/gen/installer/azure.py", line 283, in get_download_url
error	07-Sep-2016 12:51:19	    raise RuntimeError("No azure section in storage configuration")
error	07-Sep-2016 12:51:19	RuntimeError: No azure section in storage configuration
```

aaronjwood commented:
@cmaloney @orsenthil does this look good to you guys now?

orsenthil commented:
@aaronjwood  - The other day I checked out and tried the local build and it did not prove successful to me (it was not related to any of the DCOS components, but random stalling at various occasions). Our CI system exercises this on Linux and does not really give the correct picture. With your changes, are you able to consistently build on Mac? Please share that update with us. On my part, I have encountered various tricky docker-related bugs in the past, and I want to make sure  have a stable DCOS build on Mac.  

Please share the user-testing output on the section above and I will get a try again too.

aaronjwood commented:
@orsenthil yeah, I had gotten it to build all the way through before I had removed the copying of the `python3.4m` file. I haven't tried again since I reverted that change because I'll need to switch to Docker for Mac (I'm still on Docker Machine right now). Let me switch over and give it a shot.

aaronjwood commented:
@orsenthil @cmaloney It looks like I will need to make more adjustments for this to work with Docker on Mac. While it worked for me on Docker Machine this is what I eventually run into now:

```
make[1]: Entering directory `/pkg/src/mesos/build'
make[1]: Nothing to be done for `all-am'.
make[1]: Leaving directory `/pkg/src/mesos/build'
Making all in 3rdparty
make[1]: Entering directory `/pkg/src/mesos/build/3rdparty'
gzip -d -c /pkg/src/mesos/3rdparty/http-parser-2.6.2.tar.gz | tar xf -
gzip -d -c /pkg/src/mesos/3rdparty/gmock-1.7.0.tar.gz | tar xf -
test ! -e /pkg/src/mesos/3rdparty/http-parser-2.6.2.patch || patch -d http-parser-2.6.2 -p1 </pkg/src/mesos/3rdparty/http-parser-2.6.2.patch
tar: gmock-1.7.0/CHANGES: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/CMakeLists.txt: Cannot change ownership to uid 501, gid 0: Permission denied
patching file http_parser.h
touch http-parser-2.6.2-stamp
tar: gmock-1.7.0/configure.ac: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/CONTRIBUTORS: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/include/gmock/gmock-actions.h: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/include/gmock/gmock-cardinalities.h: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/include/gmock/gmock-generated-actions.h: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/include/gmock/gmock-generated-actions.h.pump: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/include/gmock/gmock-generated-function-mockers.h: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/include/gmock/gmock-generated-function-mockers.h.pump: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/include/gmock/gmock-generated-matchers.h: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/include/gmock/gmock-generated-matchers.h.pump: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/include/gmock/gmock-generated-nice-strict.h: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/include/gmock/gmock-generated-nice-strict.h.pump: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/include/gmock/gmock-matchers.h: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/include/gmock/gmock-more-actions.h: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/include/gmock/gmock-more-matchers.h: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/include/gmock/gmock-spec-builders.h: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/include/gmock/gmock.h: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/include/gmock/internal/gmock-generated-internal-utils.h: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/include/gmock/internal/gmock-generated-internal-utils.h.pump: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/include/gmock/internal/gmock-internal-utils.h: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/include/gmock/internal/gmock-port.h: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/LICENSE: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/make/Makefile: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/Makefile.am: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/msvc/2005/gmock.sln: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/msvc/2005/gmock.vcproj: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/msvc/2005/gmock_config.vsprops: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/msvc/2005/gmock_main.vcproj: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/msvc/2005/gmock_test.vcproj: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/msvc/2010/gmock.sln: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/msvc/2010/gmock.vcxproj: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/msvc/2010/gmock_config.props: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/msvc/2010/gmock_main.vcxproj: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/msvc/2010/gmock_test.vcxproj: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/README: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/scripts/fuse_gmock_files.py: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/scripts/generator/cpp/__init__.py: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/scripts/generator/cpp/ast.py: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/scripts/generator/cpp/gmock_class.py: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/scripts/generator/cpp/keywords.py: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/scripts/generator/cpp/tokenize.py: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/scripts/generator/cpp/utils.py: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/scripts/generator/gmock_gen.py: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/scripts/generator/LICENSE: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/scripts/generator/README: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/scripts/generator/README.cppclean: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/scripts/gmock-config.in: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/src/gmock-all.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/src/gmock-cardinalities.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/src/gmock-internal-utils.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/src/gmock-matchers.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/src/gmock-spec-builders.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/src/gmock.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/src/gmock_main.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock-actions_test.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock-cardinalities_test.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock-generated-actions_test.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock-generated-function-mockers_test.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock-generated-internal-utils_test.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock-generated-matchers_test.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock-internal-utils_test.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock-matchers_test.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock-more-actions_test.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock-nice-strict_test.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock-port_test.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock-spec-builders_test.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock_all_test.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock_ex_test.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock_leak_test.py: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock_leak_test_.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock_link2_test.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock_link_test.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock_link_test.h: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock_output_test.py: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock_output_test_.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock_output_test_golden.txt: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock_stress_test.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock_test.cc: Cannot change ownership to uid 501, gid 0: Permission denied
tar: gmock-1.7.0/test/gmock_test_utils.py: Cannot change ownership to uid 501, gid 0: Permission denied
tar: Exiting with failure status due to previous errors
make[1]: *** [gmock-1.7.0-stamp] Error 2
make[1]: Leaving directory `/pkg/src/mesos/build/3rdparty'
make: *** [all-recursive] Error 1
ERROR Building packages: docker exited non-zero: 2
```

aaronjwood commented:
https://issues.apache.org/jira/browse/MESOS-5925
https://github.com/apache/mesos/pull/153
https://github.com/apache/mesos/pull/153

Can this get picked into https://github.com/mesosphere/mesos?

aaronjwood commented:
@cmaloney Just fixed the conflicts and cleaned it up via squashing. Is this good to go for 1.9?

orsenthil commented:
Hi @aaronjwood , were you able to do a `./build_local.sh` with this change  and see the build on Mac?

I tried on Yosemite and Version 1.12.2-rc1-beta27 (build: 12496)  and landed up on this failure.


```
/opt/mesosphere/packages/openssl--e2fcb16174746e7b9b1a61e26c20697e960fa81e/include/openssl/crypto.h:130:31: fatal error: /opt/mesosphere/packages/openssl--e2fcb16174746e7b9b1a61e26c20697e960fa81e/include/openssl/opensslv.h: Transport endpoint is not connected
compilation terminated.
cc1: fatal error: ../include/curl/stdc-predef.h: Transport endpoint is not connected
compilation terminated.
make[2]: *** [libcurl_la-rtsp.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [libcurl_la-warnless.lo] Error 1
make[2]: *** [libcurl_la-curl_threads.lo] Error 1
make[2]: *** [libcurl_la-smtp.lo] Error 1
make[2]: *** [libcurl_la-pop3.lo] Error 1
make[2]: *** [libcurl_la-hmac.lo] Error 1
make[2]: *** [libcurl_la-pingpong.lo] Error 1
make[2]: Leaving directory `/pkg/src/curl/lib'
make[2]: *** [libcurl_la-imap.lo] Error 1
make[1]: Leaving directory `/pkg/src/curl/lib'
make[1]: *** [all] Error 2
make: *** [all-recursive] Error 1
ERROR Building packages: docker exited non-zero: 2
```

aaronjwood commented:
Hey @orsenthil, I was before I had rebased. The only thing I ran into was the issue from the 3rdparty tars in Mesos (links posted above). Let me confirm that it still works for me...

aaronjwood commented:
@orsenthil I didn't run into your issue but I'm still hitting the same tar permission issues since that fix isn't in place yet. Your Docker VM isn't out of space, is it?

orsenthil commented:
@aaronjwood , that was a spurious error. I no longer run into the Transport Endpoint not connected issue.

I do run into the tar file issue, which I see mesos is addressing (https://github.com/apache/mesos/pull/158) and then we will have complete build on Mac.

LGTM for this pull request.

Thank you!

aaronjwood commented:
Awesome, thanks!

aaronjwood commented:
Also, this is an issue I've run into with build Apache Mesos in a container using Docker for Mac https://github.com/docker/for-mac/issues/495
It should be fixed but it's not in the latest stable version yet...

aaronjwood commented:
I updated my Docker for Mac today and found that the inotify issue is fixed. Got all the way through building Mesos.

aaronjwood commented:
Should I close this out? It's been open for a long time and I don't want to keep resolving conflicts if it's something that's not being taken.

asridharan commented:
@orsenthil @mellenburg @lingmann could you guys take another look at this PR by @aaronjwood . Its been sitting here for a while. Could be potentially useful to the community??

orsenthil commented:
@aaronjwood  - Apologies for the lack of communication on this ticket. Let's close this. We discussed this multiple times and we want to avoid any potential confusion for supporting OSX as a build / development environment for DCOS.

On OSX, we build dcos on small (ubuntu) docker container, and that has been sufficient and useful to build and test DCOS locally. I will add a README note on how to accomplish under a section called (how to build on OSX). I think, the goal of this ticket will be accomplished by a README note and we will not need the changes in this PR.

Thank you!

aaronjwood commented:
@orsenthil no worries, thanks for confirming.
